### ISSUE TITLE
Cleanup `AstcodeEncoder` (and possibly other classes)

### ISSUE THREAD

**Issue by janvrany** [powerlang/powerlang#48]
There's also no clear separation of concerns and mixup of concepts. For example, there's class AstcodeEncoder`. What does it do? One might expect that it takes AST nodes and just encode it to ome representation. But no, it also does part of the compilation, see `AstcodeEncoder >> #encodeOptimizedMessage:`. And if you look inside, look like pieces of VSE-style bytecode are creeping back into structure that is supposed to be encoded AST. See for example `AstcodeEncoder >> #encodeAWithLiteral:`

**Comment by melkyades** [comment]
> There's also no clear separation of concerns and mixup of concepts. For example, there's class `AstcodeEncoder`. What does it do? One might expect that it takes AST nodes and just encode it to ome representation. But no, it also does part of the compilation, see `AstcodeEncoder >> #encodeOptimizedMessage:`. 

I agree, there's a mixture here. The problem is the following: encoding only the Smalltalk ast leaves a lot of work to the baseline JIT, which is supposed to be fast. So the encoder adds extra information about registers to make baseline JIT job easier. This work could be done as an optional separate step.
